### PR TITLE
Fix undercounting short hints

### LINKED ISSUE #3242: Sometimes, pressing the first key of a two key hint follows it prematurely
I've noticed this bug for a while, but I didn't put the time to reproduce it properly.

Copy this into a file (test.html):

```
<body>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
<a href="url">abc</a>
</body>
```

Then open the file in a `--temp-basedir` and hint the links:

2017-11-04-161855_1920x1080_scrot

pressing `l` follows the hint at this point. Another interesting thing to note is that there's only one hint starting with `l` (`la`) (shouldn't the hint display `l` in this case)?

I'm happy to take a look at this later on.

```
qutebrowser v1.0.3
Git commit: af93c0fbc (2017-11-04 17:15:20 +0100)
Backend: QtWebEngine (Chromium 49.0.2623.111)

CPython: 3.5.3
Qt: 5.7.1
PyQt: 5.7

sip: 4.18.1
colorama: no
pypeg2: 2.15
jinja2: 2.8
pygments: 2.2.0
yaml: 3.12
cssutils: noqutebrowser v1.0.3
Git commit: af93c0fbc (2017-11-04 17:15:20 +0100)
Backend: QtWebEngine (Chromium 49.0.2623.111)

CPython: 3.5.3
Qt: 5.7.1
PyQt: 5.7

sip: 4.18.1
colorama: no
pypeg2: 2.15
jinja2: 2.8
pygments: 2.2.0
yaml: 3.12
cssutils: no
attr: 16.3.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.16.2
QtNetwork SSL: OpenSSL 1.0.2l  25 May 2017

Style: QFusionStyle
Platform: Linux-4.9.0-4-amd64-x86_64-with-debian-9.2, 64bit
Linux distribution: Debian GNU/Linux 9 (stretch) (debian)
Frozen: False
Imported from /home/jay/Code/qutebrowser/qutebrowser
Qt library executable path: /usr/lib/x86_64-linux-gnu/qt5/libexec, data path: /usr/share/qt5

Paths:
cache: /tmp/qutebrowser-basedir-jd9b2psn/cache
config: /tmp/qutebrowser-basedir-jd9b2psn/config
data: /tmp/qutebrowser-basedir-jd9b2psn/data
runtime: /tmp/qutebrowser-basedir-jd9b2psn/runtime
attr: 16.3.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.16.2
QtNetwork SSL: OpenSSL 1.0.2l  25 May 2017

Style: QFusionStyle
Platform: Linux-4.9.0-4-amd64-x86_64-with-debian-9.2, 64bit
Linux distribution: Debian GNU/Linux 9 (stretch) (debian)
Frozen: False
Imported from /home/jay/Code/qutebrowser/qutebrowser
Qt library executable path: /usr/lib/x86_64-linux-gnu/qt5/libexec, data path: /usr/share/qt5

Paths:
cache: /tmp/qutebrowser-basedir-jd9b2psn/cache
config: /tmp/qutebrowser-basedir-jd9b2psn/config
data: /tmp/qutebrowser-basedir-jd9b2psn/data
runtime: /tmp/qutebrowser-basedir-jd9b2psn/runtime
```

### PR DESCRIPTION
**Pull request by jgkamat**
Possible fix for https://github.com/qutebrowser/qutebrowser/issues/3242, although I would prefer another solution if someone can come up with one.

I've manually tested the edge cases, but I would like to add unit tests as well. Is it possible for me to contstruct a fake `HintManager` in a way that I can still call the `_hin_scattered` function? When I tried to construct one manually, I got errors for referring to a non-existent window.

Closes #3242

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3329)

### CODE REVIEW

**Comment by The-Compiler** [comment]
I think this solution is okay - it's probably more readable than whatever else we could come up with. I can't think of anything else either, I might ask some other people today if I find the time.

As for writing unit tests, you need to use a few fixtures to set dependencies up. This works for me:

```python
def test_hint_manager(win_registry, mode_manager):
    man = hints.HintManager(0, 0)
```

File: tests/end2end/test_hints_html.py [hunk 1]
```diff
@@ -22,12 +22,17 @@
...
+import string
+import functools
+import operator
 
 import attr
 import yaml
 import pytest
 import bs4
 
+import qutebrowser.browser.hints
```
Comment (The-Compiler) [inline tests/end2end/test_hints_html.py]: FWIW `from qutebrowser.browser import hints` is okay - it's just things like `from qutebrowser.browser.hints import HintManager` I try to avoid.

File: tests/end2end/test_hints_html.py [hunk 2]
```diff
@@ -146,3 +151,45 @@ def test_word_hints_issue1393(quteproc, tmpdir):
...
         quteproc.send_cmd(':follow-hint {}'.format(hint))
         quteproc.wait_for_load_finished('data/{}'.format(target))
+
+
+@pytest.mark.parametrize('min_len', [0, 3])
+@pytest.mark.parametrize('num_chars', [9])
+@pytest.mark.parametrize('num_elements', range(1, 26))
+def test_scattered_hints_count(win_registry, mode_manager, min_len,
+                               num_chars, num_elements):
+    """Test scattered hints function."""
```
Comment (The-Compiler) [inline tests/end2end/test_hints_html.py]: Maybe explain what we're testing for here exactly?

File: tests/end2end/test_hints_html.py [hunk 3]
```diff
@@ -146,3 +151,45 @@ def test_word_hints_issue1393(quteproc, tmpdir):
...
+    # pylint: disable=W0141
```
Comment (The-Compiler) [inline tests/end2end/test_hints_html.py]: Please use the name instead of the error code here.

File: tests/end2end/test_hints_html.py [hunk 4]
```diff
@@ -146,3 +151,45 @@ def test_word_hints_issue1393(quteproc, tmpdir):
...
+    chars = string.ascii_lowercase[:num_chars]
+
+    hints = manager._hint_scattered(min_len, chars,
+                                    list(range(num_elements)))
+
+    # Check if hints are unique
+    assert len(hints) == len(set(hints))
+
+    # Check if any hints are shorter than min_len
+    assert not any([x for x in hints if len(x) < min_len])
```
Comment (The-Compiler) [inline tests/end2end/test_hints_html.py]: You can drop the `[]` here.

File: tests/end2end/test_hints_html.py [hunk 5]
```diff
@@ -146,3 +151,45 @@ def test_word_hints_issue1393(quteproc, tmpdir):
...
+
+    # Check we don't have more than 2 link lengths
+    # Eg: 'a' 'bc' and 'def' cannot be in the same hint string
+    hint_lens = set(map(len, hints))
```
Comment (The-Compiler) [inline tests/end2end/test_hints_html.py]: I'd prefer `{len(h) for h in hints}`

File: tests/end2end/test_hints_html.py [hunk 6]
```diff
@@ -146,3 +151,45 @@ def test_word_hints_issue1393(quteproc, tmpdir):
...
+    longest_hints = filter(lambda x: len(x) == max(hint_lens), hints)
+
+    if min_len < max(hint_lens) - 1:
+        # Check if we have any unique prefixes. For example, 'la'
+        # alone, with no other 'l<x>'
+        count_map = {}
+        for x in longest_hints:
+            prefix = x[:-1]
+            count_map[prefix] = count_map.get(prefix, 0) + 1
+        assert not any(filter(lambda x: x == 1, count_map.values()))
```
Comment (The-Compiler) [inline tests/end2end/test_hints_html.py]: I don't understand this - isn't this basically the equivalent of `any([1, 1, 1])` which will always be `True`, i.e. you're just using `any()` to see whether a list is empty?

This would be much clearer IMHO: `assert not any(e == 1 for e in count_map.values())`
or even better: `assert all(e != 1 for e in count_map.values())`

File: tests/end2end/test_hints_html.py (hunk 1 above)
Comment (jgkamat) [reply tests/end2end/test_hints_html.py]: Ah, that makes a lot more sense now, so modules are fine :)

In this specific case, I'd prefer to leave it as-is, as I don't like taking up the name 'hints' in a hint testing file (and I'm only using it once anyway).

File: qutebrowser/browser/hints.py [hunk 7]
```diff
@@ -446,8 +446,17 @@ def _hint_scattered(self, min_chars, chars, elems):
...
         # (needed - 1) digits in length.
         if needed > min_chars:
-            short_count = math.floor((len(chars) ** needed - len(elems)) /
-                                     len(chars))
+            total_space = len(chars) ** needed
+            # Calculate short_count naively, by finding the avaiable space and
+            # dividing by the number of spots we would loose by adding a
+            # short element
+            short_count = math.floor((total_space - len(elems)) /
+                                     (len(chars)))
```
Comment (The-Compiler) [inline qutebrowser/browser/hints.py]: There are unneeded parens here, but I'll fix it up when merging.

File: tests/end2end/test_hints_html.py [hunk 8]
```diff
@@ -146,3 +151,51 @@ def test_word_hints_issue1393(quteproc, tmpdir):
         quteproc.wait_for(message='hints: *', category='hints')
         quteproc.send_cmd(':follow-hint {}'.format(hint))
         quteproc.wait_for_load_finished('data/{}'.format(target))
+
+
+@pytest.mark.parametrize('min_len', [0, 3])
```
Comment (The-Compiler) [inline tests/end2end/test_hints_html.py]: This isn't really the right place for a test like this (it should be in a new `test/unit/browser/test_hints.py`), but I'll move it while merging.

File: qutebrowser/browser/hints.py (hunk 7 above)
Comment (The-Compiler) [reply qutebrowser/browser/hints.py]: Done in 81045fb1bd88a2dc9c6136ee130d54929d751881

File: tests/end2end/test_hints_html.py (hunk 8 above)
Comment (The-Compiler) [reply tests/end2end/test_hints_html.py]: Done in af638ec430be25c5f61c697a49efaeb7cade2bd6
